### PR TITLE
Implement the MultiStateWeightProvider and refactor the WeightProvider

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
The `MultiStateWeightProvider` is a provider that will allow us to perform gain scheduling. The user should provide a configuration file containing the name and the weight for each state. Then the provider will interpolate among the states using a FirstOrder stable system. The settling time of the system should be set in the configuration file.

I also implemented the associated tests and the python bindings.

I also add the possibility to set the weight provider directly in the `ILinearTaskSolver::addTask()` method.

The PR also introduces the concept of output and input ports. These two interfaces contain functions to get and set signals only. And advanceable inherits from them. Thanks to this w can split the concept of advanceable from the concept of getting values. For instance
```cpp
class Foo : public Advanceable<double> {...};

void f(const OutputPort<double>& port);


int main() {
    Foo foo;
    f(foo);
```

Now the user knows that `f` will only get the output of the advanceable and nothing more, (for instance it will never call `advance` or `setInput`) 

---

This substitutes #507

### CODE REVIEW

File: src/System/include/BipedalLocomotion/System/ConstantWeightProvider.h [hunk 1]
```diff
@@ -41,22 +40,34 @@ struct ConstantWeightProvider : public IWeightProvider
...
     /**
      * Initialize constant weight provider.
      * @param handler pointer to the parameter handler.
      * @note The following parameters are required:
-     * |  Parameter Name  |        Type      |                          Description                              | Mandatory |
+     * |  Parameter Name  |        Type      |                          Description | Mandatory |
      * |:----------------:|:----------------:|:-----------------------------------------------------------------:|:---------:|
-     * |    `weight`      | `vector<double>` |  Vector representing the diagonal matrix of a constant weight     |    Yes    |
+     * |    `weight`      | `vector<double>` |  Vector representing the diagonal matrix of a
+     * constant weight     |    Yes    |
```
Comment (S-Dafarra) [inline src/System/include/BipedalLocomotion/System/ConstantWeightProvider.h:67]: I think that this comment got automatically cut in two lines but it probably shouldn't.

File: src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h [hunk 2]
```diff
@@ -48,14 +51,30 @@ template <class _Task, class _State> class ILinearTaskSolver : public Source<_St
      * @param taskName unique name associated to the task.
      * @param priority Priority associated to the task. The lower the number the higher the
      * priority.
-     * @param weight Weight associated to the task. This parameter is optional. The default value is
-     * an object that does not contain any value. So is an invalid weight.
-     * @return true if the task has been added to the inverse kinematics.
+     * @param weightProvider Weight provider associated to the task. This parameter is optional. The
+     * default value is an object that does not contain any value. So is an invalid provider.
```
Comment (S-Dafarra) [inline src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h:55]: As for ``QPTSID``, maybe it is worth explaining a bit what it means to set an invalid provider

File: src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h [hunk 3]
```diff
@@ -48,14 +51,30 @@ template <class _Task, class _State> class ILinearTaskSolver : public Source<_St
...
+
+    /**
+     * Add a linear task in the solver.
+     * @param task pointer to a given linear task
+     * @param taskName unique name associated to the task.
+     * @param priority Priority associated to the task. The lower the number the higher the
+     * priority.
+     * @param weight Weight associated to the task.
+     * @return true if the task has been added to the solver.
+     * @note If this method is used the solver assume that the weight is a constant value.
```
Comment (S-Dafarra) [inline src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h:72]:
```suggestion
     * @note The solver assumes the weight is a constant value.
```

File: src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h [hunk 4]
```diff
@@ -64,8 +83,18 @@ template <class _Task, class _State> class ILinearTaskSolver : public Source<_St
...
-    virtual bool setTaskWeightProvider(const std::string& taskName,
-                                       std::shared_ptr<const IWeightProvider> weightProvider)
+    virtual bool setTaskWeight(const std::string& taskName,
+                               std::shared_ptr<const WeightProviderPort> weightProvider)
+        = 0;
+
+    /**
+     * Set the weight associated to an already existing task
+     * @param taskName name associated to the task
+     * @param weight new Weight associated to the task. A constant weight is assumed.
```
Comment (S-Dafarra) [inline src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h:93]:
```suggestion
     * @param weight new weight associated to the task. A constant weight is assumed.
```

File: src/IK/include/BipedalLocomotion/IK/QPInverseKinematics.h [hunk 5]
```diff
@@ -59,40 +58,65 @@ class QPInverseKinematics : public IntegrationBasedIK
...
 
     /**
-     * Add a linear task in the inverse kinematics
+     * Add a linear task in the solver.
+     * @param task pointer to a given linear task
+     * @param taskName unique name associated to the task.
+     * @param priority Priority associated to the task. The lower the number the higher the
+     * priority.
+     * @param weightProvider Weight provider associated to the task. This parameter is optional. The
+     * default value is an object that does not contain any value. So is an invalid provider.
```
Comment (S-Dafarra) [inline src/IK/include/BipedalLocomotion/IK/QPInverseKinematics.h:67]: Same as in the other comments

File: src/IK/include/BipedalLocomotion/IK/QPInverseKinematics.h [hunk 6]
```diff
@@ -59,40 +58,65 @@ class QPInverseKinematics : public IntegrationBasedIK
...
-    bool
-    setTaskWeightProvider(const std::string& taskName,
-                          std::shared_ptr<const System::IWeightProvider> weightProvider) override;
+    bool setTaskWeight(const std::string& taskName,
+                       std::shared_ptr<const System::WeightProviderPort> weightProvider) override;
+
+    /**
+     * Set the weight associated to an already existing task
+     * @param taskName name associated to the task
+     * @param weight new Weight associated to the task. A constant weight is assumed.
```
Comment (S-Dafarra) [inline src/IK/include/BipedalLocomotion/IK/QPInverseKinematics.h:107]:
```suggestion
     * @param weight new weight associated to the task. A constant weight is assumed.
```

File: src/IK/src/QPInverseKinematics.cpp [hunk 7]
```diff
@@ -192,7 +203,7 @@ bool QPInverseKinematics::addTask(std::shared_ptr<QPInverseKinematics::Task> tas
         }
 
         // add the weight In this case we assume that the weight will be constant
-        m_pimpl->tasks[taskName].weightProvider = std::make_shared<System::ConstantWeightProvider>(weight.value());
+        m_pimpl->tasks[taskName].weightProvider = weightProvider;
```
Comment (S-Dafarra) [inline src/IK/src/QPInverseKinematics.cpp:206]: As in the other comment

File: src/IK/src/QPInverseKinematics.cpp [hunk 8]
```diff
@@ -174,16 +185,16 @@ bool QPInverseKinematics::addTask(std::shared_ptr<QPInverseKinematics::Task> tas
 
     // IF a weight the priority is 1 and the weight is provided means that the user wants to specify
     // a constant weight in the task
-    if (priority == 1 && weight)
+    if (priority == 1 && weightProvider)
```
Comment (S-Dafarra) [inline src/IK/src/QPInverseKinematics.cpp:188]: As in the other comment

**Review by S-Dafarra — COMMENTED** [review]
Mostly comments about documentation

**Comment by GiulioRomualdi** [comment]
@S-Dafarra I think I handled all your comments

**Comment by GiulioRomualdi** [comment]
brew fails as already reported  in: https://github.com/robotology/robotology-superbuild/issues/1234

**Review by S-Dafarra — APPROVED** [review]

**Pull request merged**